design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,119 @@
+# Overview
+VM Cloning is the action of creating a new VM from an existing one that has the exact same setting. This includes
+VM disks, network definition, CPU topology and so on.
+
+## Motivation
+The ability to clone a VM is a generic virtualization operation which may be useful for several use-cases.
+
+For example, a system admin / user might want to base a new VM on an existing one. In such a case we can clone a VM and
+then change it according to our needs. In addition, a cluster admin might want to try different configurations
+(for example, to check whether other configuration provide better performance for his workload), in this case he can
+clone the existing VM, test it and decide whether to move the new settings.
+
+Naively copy/pasting a VM spec would not work in most cases therefore this serves as an automation for cloning safely.
+
+## Goals
+Have a mechanism to clone VMs. Original and cloned VM would have the same:
+* Disks
+* Network configuration
+* Volumes
+* CPU topology
+
+Also, provide an ability to fine-tune what's being cloned. For example, it is possible to clone or skip
+annotations / labels.  
+
+## Non Goals
+In first implementation iteration the following limitations exist:
+* Only offline cloning will be supported, meaning cloning a VM when it's stopped.
+* Cloned VM would be in the same namespace as the original one.
+* Host disks would not be supported.
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
```

Comment:
As having a new CRD introduces a higher level of complexity to the system, could you please share why this approach is preferred over the same logic to exist in some client/library (`virtctl`)?

BTW, unlike the migration CRD, this one does not define any status fields (i.e. success, failure and whatever other data is relevant).

Reply:
I think that using CRDs has some benefits:
- It is the "Kubernetes way" of performing such operations. It is declarative, easy to use and similar to other Kubernetes operations.
- Implementing all logic in `virtctl` implies that all of the cloning operation is performed by the client side. In general this is less robust, since the client can, for example, disconnect from the server at the middle of the cloning operation. This cannot happen if the cloning request resides in etcd and controllers watch this resource to perform the operation.
- We can have status in this CRD as well
- Some virtctl operations are actually implemented with a CRD, `virtctl migrate` for example.

> As having a new CRD introduces a higher level of complexity to the system

What is the big disadvantage here in your opinion? What are the benefits in avoiding a CRD here?

Reply:
the clone operation needs to be managed by the backend in a declarative way that all client side tooling can invoke. so, virtctl, kubectl, and something like the OCP UI shoudl all be able to invoke a clone operation using the same backend logic.

Reply:
> Regarding being cut in the middle of the operation, that is true also when defining any other object (e.g. the VMI). But I may be missing something here.

That's right. If, for example, the creation of a VMI object would be done from `virtctl` (i.e. the client side) and would consist of say 5 steps - if the client had crashed after the 3rd step we would be stuck in an inconsistent state. This cannot happen when a VMI CR is posted to the cluster since even if some of the components crash, as soon as they are back up they would read etcd, understand that a VMI creation is stuck and what's needed to be done to complete the creation operation.

Here's a better phrasing than "kubernetes way": it is Kubernetes' solution to these kind of problems, which is an important capability.

> Isn't the cloning itself limited to just copying specific specs with some limitations?

Firstly, we're not sure what would happen in the future. Secondly, even if we just copy the spec with some limitations / modifications, we may trigger complex operations (like DV cloning). If these operations fail, in the future we may retry, fallback to different solutions, etc'.

> I think that a CRD is not cheap, it adds a great maintenance burden and should be used carefully

As I previously asked: `What is the big disadvantage here in your opinion? What are the benefits in avoiding a CRD here?`. I'm still not sure why CRD is not cheap and where is the disadvantage you're speaking about. Why do you refer it as "big guns"?

Reply:
As I see it,having a CRD representing consistent & core objects like  Kubevirt cluster config, VM, VMI, etc.. makes it a great fit.
But, when you start using on a per-feature granularity, especially imperative operations, it becomes an overkill mechanism which bloats one product with too many API extensions, which are challenging to maintain.
CRD/s are more expensive to maintain because they require controllers which need to monitor them, change them, deploy, remove them, etc. It is also costly from the users side, as they need to know about them and browse through potential a large number of them in the system.

AFAIK, we already have several imperative API commands (e.g. hotplug for storage, and soon for network) which do not use CRD/s. They currently use a subresource on the virt-api entrypoint. There are also domain operations like reboot, shutdown, etc.
You could also argue that they have a flow that may be implemented using a CRD, for the same reasoning you mentioned before (e.g. being interrupted in the middle, fitting the k8s way).

I am trying to understand if there are other alternatives to extend the functionality of kubevirt without adding CRD/s. For the long run, I think it is worth exploring other options to make sure we do not overuse a k8s option, increasing kubevirt complexity beyond what is a must.
This design suggestion is presenting only one option, which makes it hard to see that other things have been explored. I think the effort to find other alternative is worth it, even if eventual a CRD will be the final choice.

To be specific, I am interested to understand why an imperative API through a subresource and potentially an annotation (in case state needs to be saved somewhere) is not enough in this case. Can't a single command on the API generate the needed VM spec and other definitions to clone the VM?

Reply:
> To be specific, I am interested to understand why an imperative API through a subresource and potentially an annotation (in case state needs to be saved somewhere) is not enough in this case. Can't a single command on the API generate the needed VM spec and other definitions to clone the VM?

What might be problematic here IMO is:
* Complexity of arguments: if we have a large number of arguments and options, using the command line can be messy. This is especially true when providing lists of, e.g., annotations/labels/network interfaces to include/exclude. 
* We're not sure where this feature will lead us. It might be possible to clone VMIs in the future, for example. It might be limiting to have a single subresource (e.g. VM's subresource) to handle this.

In your examples, e.g. booting, restarting, migrating, etc, we don't have such problems since the operations barely have arguments and are relatively closed to expansions.

----

Another thing is that I'm not sure I fundamentally agree with the assumption that CRDs bring more complexity.

**From the performance / runtime perspective:** It's true that it's more expensive to maintain CRDs because we bring new controllers to watch and handle them, but this price certainly brings value:
* Controllers are much more reliable, robust and scalable then `virtctl`
* For developers, it's easier to maintain since controllers are Kubernetes' idiomatic way of working. `virtctl` is something different which brings more maintainability complexity from the developers' perspective.
* Controllers are optimized for distributed systems. The whole mechanism exposes a lot of knobs to fine-tune the resource intensity of each controller.

**From the performance / runtime perspective:** for similar reasons I'm not sure that `virtctl` is easier to understand or use. You mentioned that `It is also costly from the users side, as they need to know about them and browse through potential a large number of them in the system` - but the user would need to browse a large number of options either way. Why do you think that browsing on `virtctl`, a custom binary, would be easier than browsing on CRDs, API that every Kubernetes user is familiar with?

To summarize: I agree that the "Kubernetes way" is not holy and we should always consider alternatives. On the other hand, I think that with any platform there are many automatic benefits gained by doing things in the "platform way". In other words, there should be a good enough reason not to go that way. In this specific case I don't see a huge value from not exposing a CRD.

Reply:
> AFAIK, we already have several imperative API commands (e.g. hotplug for storage, and soon for network) which do not use CRDs

hotplug is declarative.  it uses the VM/VMI status to make a volume request which models the declarative workflow for us.

> I am trying to understand if there are other alternatives to extend the functionality of kubevirt without adding CRD/s.

It you want to model a declarative workflow, using a new or existing CRD is the way to do it in k8s if you're a dev on the project itself. Something like a tekton pipeline would be the way to model it as a devops engineer. 

I'd prefer we keep the clone workflow outside of the VM object itself, similar to how we kept the snapshot/restore apis separate from the VM object. There's existing precedence within our ecosystem to support the pattern of using a CRD to model the clone workflow.